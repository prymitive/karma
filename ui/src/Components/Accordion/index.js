import React from "react";

import Collapsible from "react-collapsible";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronUp } from "@fortawesome/free-solid-svg-icons/faChevronUp";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";

const Trigger = ({ text, isOpen }) => (
  <div className="d-flex flex-row justify-content-between">
    <div>{text}</div>
    <div>
      <FontAwesomeIcon
        icon={isOpen ? faChevronDown : faChevronUp}
        className="text-muted"
      />
    </div>
  </div>
);

const Accordion = ({ text, content, extraProps }) => (
  <Collapsible
    triggerTagName="div"
    transitionTime={50}
    className="card"
    openedClassName="card"
    triggerClassName="card-header cursor-pointer"
    triggerOpenedClassName="card-header cursor-pointer bg-light"
    contentOuterClassName="collapse show"
    contentInnerClassName="card-body my-2"
    {...extraProps}
    trigger={<Trigger text={text} isOpen={false} />}
    triggerWhenOpen={<Trigger text={text} isOpen={true} />}
  >
    {content}
  </Collapsible>
);

export { Accordion };
