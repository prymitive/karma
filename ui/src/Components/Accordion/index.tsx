import { FC, ReactNode, useState } from "react";

export const AccordionItem: FC<{
  text: string;
  content: ReactNode;
  defaultIsOpen?: boolean;
}> = ({ text, content, defaultIsOpen }) => {
  const [isOpen, setIsOpen] = useState<boolean>(defaultIsOpen || false);

  return (
    <div className="accordion-item">
      <h2 className="accordion-header">
        <button
          className={`accordion-button ${isOpen ? "" : "collapsed"}`}
          type="button"
          onClick={() => setIsOpen((val) => !val)}
        >
          {text}
        </button>
      </h2>
      <div className={`accordion-collapse ${isOpen ? "show" : ""}`}>
        {isOpen ? <div className="accordion-body">{content}</div> : null}
      </div>
    </div>
  );
};

export const Accordion: FC = ({ children }) => {
  return <div className="accordion">{children}</div>;
};
