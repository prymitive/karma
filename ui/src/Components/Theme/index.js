import React, { useEffect } from "react";
import ReactDOM from "react-dom";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSun } from "@fortawesome/free-solid-svg-icons/faSun";

const DarkTheme = React.lazy(() => import("Styles/DarkTheme"));
const LightTheme = React.lazy(() => import("Styles/LightTheme"));

const Placeholder = () => {
  return ReactDOM.createPortal(
    <div
      style={{
        zIndex: 2000,
        backgroundColor: "#455a64",
        position: "fixed",
        top: 0,
        left: 0,
        width: "100vw",
        height: "100vh",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FontAwesomeIcon
        icon={faSun}
        size="lg"
        spin
        style={{
          color: "#5e7a88",
          fontSize: "14rem",
        }}
      />
    </div>,
    document.body
  );
};

const ThemeContext = React.createContext();

const BodyTheme = () => {
  const context = React.useContext(ThemeContext);

  useEffect(() => {
    document.body.classList.toggle("theme-dark", context.isDark);
    document.body.classList.toggle("theme-light", !context.isDark);
  }, [context.isDark]);

  return (
    <React.Suspense fallback={<Placeholder />}>
      {context.isDark ? <DarkTheme /> : <LightTheme />}
    </React.Suspense>
  );
};

export { BodyTheme, ThemeContext };
