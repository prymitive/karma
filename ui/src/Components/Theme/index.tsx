import React, { FC, useEffect } from "react";
import ReactDOM from "react-dom";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSun } from "@fortawesome/free-solid-svg-icons/faSun";

import { Styles } from "react-select";

const DarkTheme = React.lazy(() => import("Styles/DarkTheme"));
const LightTheme = React.lazy(() => import("Styles/LightTheme"));

const Placeholder = () => {
  return ReactDOM.createPortal(
    <div
      style={{
        zIndex: 2000,
        backgroundColor: "#455a64",
        position: "fixed",
        top: 0,
        left: 0,
        width: "100vw",
        height: "100vh",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FontAwesomeIcon
        icon={faSun}
        size="lg"
        spin
        style={{
          color: "#5e7a88",
          fontSize: "14rem",
        }}
      />
    </div>,
    document.body
  );
};

export interface ThemeCtx {
  isDark: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  reactSelectStyles: Styles<any, any>;
  animations: {
    duration: number;
  };
}

const ThemeContext = React.createContext({
  isDark: false,
  reactSelectStyles: {},
  animations: { duration: 1000 },
} as ThemeCtx);

const BodyTheme: FC = () => {
  const context = React.useContext(ThemeContext);

  useEffect(() => {
    document.body.classList.toggle("theme-dark", context.isDark);
    document.body.classList.toggle("theme-light", !context.isDark);
  }, [context.isDark]);

  return (
    <React.Suspense fallback={<Placeholder />}>
      {context.isDark ? <DarkTheme /> : <LightTheme />}
    </React.Suspense>
  );
};

export { BodyTheme, ThemeContext };
