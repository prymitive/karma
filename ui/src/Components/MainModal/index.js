import React, { useState, useCallback } from "react";
import PropTypes from "prop-types";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCog } from "@fortawesome/free-solid-svg-icons/faCog";
import { faSpinner } from "@fortawesome/free-solid-svg-icons/faSpinner";

import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { TooltipWrapper } from "Components/TooltipWrapper";
import { Modal } from "Components/Modal";

// https://github.com/facebook/react/issues/14603
const MainModalContent = React.lazy(() =>
  import("./MainModalContent").then((module) => ({
    default: module.MainModalContent,
  }))
);

const MainModal = ({ alertStore, settingsStore }) => {
  const [isVisible, setIsVisible] = useState(false);

  const toggle = useCallback(() => setIsVisible(!isVisible), [isVisible]);

  return (
    <React.Fragment>
      <li
        className={`nav-item components-navbar-button ${
          isVisible ? "border-info" : ""
        }`}
      >
        <TooltipWrapper title="Settings">
          <span
            id="components-settings"
            className="nav-link cursor-pointer"
            onClick={toggle}
          >
            <FontAwesomeIcon icon={faCog} />
          </span>
        </TooltipWrapper>
      </li>
      <Modal isOpen={isVisible} toggleOpen={toggle}>
        <React.Suspense
          fallback={
            <h1 className="display-1 text-placeholder p-5 m-auto">
              <FontAwesomeIcon icon={faSpinner} size="lg" spin />
            </h1>
          }
        >
          <MainModalContent
            alertStore={alertStore}
            settingsStore={settingsStore}
            onHide={() => setIsVisible(false)}
            isVisible={isVisible}
            expandAllOptions={false}
          />
        </React.Suspense>
      </Modal>
    </React.Fragment>
  );
};
MainModal.propTypes = {
  alertStore: PropTypes.instanceOf(AlertStore).isRequired,
  settingsStore: PropTypes.instanceOf(Settings).isRequired,
};

export { MainModal };
