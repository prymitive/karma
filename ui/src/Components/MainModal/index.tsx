import React, { FC, useState, useCallback } from "react";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSlidersH } from "@fortawesome/free-solid-svg-icons/faSlidersH";
import { faSpinner } from "@fortawesome/free-solid-svg-icons/faSpinner";

import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { TooltipWrapper } from "Components/TooltipWrapper";
import { Modal } from "Components/Modal";

// https://github.com/facebook/react/issues/14603
const MainModalContent = React.lazy(() =>
  import("./MainModalContent").then((module) => ({
    default: module.MainModalContent,
  }))
);

const MainModal: FC<{
  alertStore: AlertStore;
  settingsStore: Settings;
}> = ({ alertStore, settingsStore }) => {
  const [isVisible, setIsVisible] = useState<boolean>(false);

  const toggle = useCallback(() => setIsVisible(!isVisible), [isVisible]);

  return (
    <>
      <li
        className={`nav-item components-navbar-button ${
          isVisible ? "border-info" : ""
        } ml-auto`}
      >
        <TooltipWrapper title="Settings">
          <span
            id="components-settings"
            className="nav-link cursor-pointer"
            onClick={toggle}
          >
            <FontAwesomeIcon icon={faSlidersH} fixedWidth />
          </span>
        </TooltipWrapper>
      </li>
      <Modal isOpen={isVisible} toggleOpen={toggle}>
        <React.Suspense
          fallback={
            <h1 className="display-1 text-placeholder p-5 m-auto">
              <FontAwesomeIcon icon={faSpinner} size="lg" spin />
            </h1>
          }
        >
          <MainModalContent
            alertStore={alertStore}
            settingsStore={settingsStore}
            onHide={() => setIsVisible(false)}
            expandAllOptions={false}
          />
        </React.Suspense>
      </Modal>
    </>
  );
};

export { MainModal };
