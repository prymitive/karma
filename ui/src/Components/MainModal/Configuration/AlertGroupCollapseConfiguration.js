import React from "react";
import PropTypes from "prop-types";

import { useObserver } from "mobx-react-lite";

import Select from "react-select";

import { Settings } from "Stores/Settings";
import { ThemeContext } from "Components/Theme";

const AlertGroupCollapseConfiguration = ({ settingsStore }) => {
  if (
    !Object.values(settingsStore.alertGroupConfig.options)
      .map((o) => o.value)
      .includes(settingsStore.alertGroupConfig.config.defaultCollapseState)
  ) {
    settingsStore.alertGroupConfig.config.defaultCollapseState =
      settingsStore.alertGroupConfig.options.collapsedOnMobile.value;
  }

  const valueToOption = (val) => {
    return {
      label: settingsStore.alertGroupConfig.options[val].label,
      value: val,
    };
  };

  const onCollapseChange = (newValue, actionMeta) => {
    settingsStore.alertGroupConfig.config.defaultCollapseState = newValue.value;
  };

  const context = React.useContext(ThemeContext);

  return useObserver(() => (
    <div className="form-group mb-0">
      <Select
        styles={context.reactSelectStyles}
        classNamePrefix="react-select"
        instanceId="configuration-collapse"
        defaultValue={valueToOption(
          settingsStore.alertGroupConfig.config.defaultCollapseState
        )}
        options={Object.values(settingsStore.alertGroupConfig.options)}
        onChange={onCollapseChange}
        hideSelectedOptions
      />
    </div>
  ));
};
AlertGroupCollapseConfiguration.propTypes = {
  settingsStore: PropTypes.instanceOf(Settings).isRequired,
};

export { AlertGroupCollapseConfiguration };
