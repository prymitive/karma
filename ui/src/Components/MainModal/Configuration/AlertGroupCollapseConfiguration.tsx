import React, { FC } from "react";

import { observer } from "mobx-react-lite";

import Select from "react-select";

import { OptionT } from "Common/Select";
import { Settings, CollapseStateT } from "Stores/Settings";
import { AnimatedMenu } from "Components/Select";
import { ThemeContext } from "Components/Theme";

const AlertGroupCollapseConfiguration: FC<{
  settingsStore: Settings;
}> = observer(({ settingsStore }) => {
  if (
    !Object.values(settingsStore.alertGroupConfig.options)
      .map((o) => o.value)
      .includes(settingsStore.alertGroupConfig.config.defaultCollapseState)
  ) {
    settingsStore.alertGroupConfig.setDefaultCollapseState("collapsedOnMobile");
  }

  const valueToOption = (val: CollapseStateT): OptionT => {
    return {
      label: settingsStore.alertGroupConfig.options[val].label,
      value: val,
    };
  };

  const onCollapseChange = (newValue: CollapseStateT) => {
    settingsStore.alertGroupConfig.setDefaultCollapseState(newValue);
  };

  const context = React.useContext(ThemeContext);

  return (
    <div className="mb-0">
      <Select
        styles={context.reactSelectStyles}
        classNamePrefix="react-select"
        instanceId="configuration-collapse"
        defaultValue={valueToOption(
          settingsStore.alertGroupConfig.config.defaultCollapseState
        )}
        options={Object.values(settingsStore.alertGroupConfig.options)}
        onChange={(option) =>
          onCollapseChange((option as OptionT).value as CollapseStateT)
        }
        hideSelectedOptions
        components={{ Menu: AnimatedMenu }}
      />
    </div>
  );
});

export { AlertGroupCollapseConfiguration };
