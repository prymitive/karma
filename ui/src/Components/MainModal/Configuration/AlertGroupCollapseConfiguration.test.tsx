import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { MockThemeContext } from "__fixtures__/Theme";
import { Settings } from "Stores/Settings";
import { ThemeContext } from "Components/Theme";
import { AlertGroupCollapseConfiguration } from "./AlertGroupCollapseConfiguration";

let settingsStore: Settings;

beforeEach(() => {
  settingsStore = new Settings(null);
});

const FakeConfiguration = () => {
  return mount(
    <AlertGroupCollapseConfiguration settingsStore={settingsStore} />,
    {
      wrappingComponent: ThemeContext.Provider,
      wrappingComponentProps: { value: MockThemeContext },
    }
  );
};

describe("<AlertGroupCollapseConfiguration />", () => {
  it("matches snapshot with default values", () => {
    const tree = FakeConfiguration();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("resets stored config to defaults if it is invalid", (done) => {
    settingsStore.alertGroupConfig.setDefaultCollapseState("foo" as any);
    const tree = FakeConfiguration();
    const select = tree.find("div.react-select__value-container");
    expect(select.text()).toBe(
      settingsStore.alertGroupConfig.options.collapsedOnMobile.label
    );
    setTimeout(() => {
      expect(settingsStore.alertGroupConfig.config.defaultCollapseState).toBe(
        settingsStore.alertGroupConfig.options.collapsedOnMobile.value
      );
      done();
    }, 200);
  });

  it("rendered correct default value", (done) => {
    settingsStore.alertGroupConfig.setDefaultCollapseState("expanded");
    const tree = FakeConfiguration();
    const select = tree.find("div.react-select__value-container");
    setTimeout(() => {
      expect(select.text()).toBe(
        settingsStore.alertGroupConfig.options.expanded.label
      );
      done();
    }, 200);
  });

  it("clicking on a label option updates settingsStore", (done) => {
    const tree = FakeConfiguration();
    tree
      .find("input#react-select-configuration-collapse-input")
      .simulate("change", { target: { value: " " } });
    const options = tree.find("div.react-select__option");
    options.at(1).simulate("click");
    setTimeout(() => {
      expect(settingsStore.alertGroupConfig.config.defaultCollapseState).toBe(
        settingsStore.alertGroupConfig.options.collapsed.value
      );
      done();
    }, 200);
  });
});
