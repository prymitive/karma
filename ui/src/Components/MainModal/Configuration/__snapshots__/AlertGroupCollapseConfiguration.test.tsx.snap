// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertGroupCollapseConfiguration /> matches snapshot with default values 1`] = `
"
<div class=\\"mb-0\\">
  <div class=\\" css-2b097c-container\\">
    <span aria-live=\\"polite\\"
          aria-atomic=\\"false\\"
          aria-relevant=\\"additions text\\"
          class=\\"css-1f43avz-a11yText-A11yText\\"
    >
    </span>
    <div class=\\"react-select__control css-cat782-control\\">
      <div class=\\"react-select__value-container react-select__value-container--has-value css-17yk2dy-ValueContainer\\">
        <div class=\\"react-select__single-value css-1wh03ml-singleValue\\">
          Collapse on mobile
        </div>
        <div class=\\"css-ps6ina-Input\\">
          <div class=\\"react-select__input\\"
               style=\\"display: inline-block;\\"
          >
            <input autocapitalize=\\"none\\"
                   autocomplete=\\"off\\"
                   autocorrect=\\"off\\"
                   id=\\"react-select-configuration-collapse-input\\"
                   spellcheck=\\"false\\"
                   tabindex=\\"0\\"
                   type=\\"text\\"
                   aria-autocomplete=\\"list\\"
                   style=\\"box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;\\"
                   value
            >
            <div style=\\"position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;\\">
            </div>
          </div>
        </div>
      </div>
      <div class=\\"react-select__indicators css-vcwr3k-IndicatorsContainer\\">
        <span class=\\"react-select__indicator-separator css-1okebmr-indicatorSeparator\\">
        </span>
        <div class=\\"react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer\\"
             aria-hidden=\\"true\\"
        >
          <svg height=\\"20\\"
               width=\\"20\\"
               viewbox=\\"0 0 20 20\\"
               aria-hidden=\\"true\\"
               focusable=\\"false\\"
               class=\\"css-tj5bde-Svg\\"
          >
            <path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\">
            </path>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
"
`;
